File: q1-25/q20.py
# 2020.07.24
# Problem Statement:
# https://leetcode.com/problems/valid-parentheses/
# I have to admit that I looked up the solution for better answer:(

class Solution:
    def isValid(self, s: str) -> bool:        
        # early check
        if len(s) % 2 != 0: return False
        
        # loop until all valid combinations are removed
        while "()" in s or "{}" in s or "[]" in s:
            s = s.replace("()", "")
            s = s.replace("{}", "")
            s = s.replace("[]", "")

        # if nothing left, then valid    
        if s == "": return True
        else: return False

File: Amazon/q20.java
// 2022.05.19 midnight
// Problem Statement:
// https://leetcode.com/problems/valid-parentheses/

// idea: use stack to push right parantheses that should appear in the reverse order of the left parantheses
// if the right parantheses found does not match the stack's top, return false
// https://leetcode.com/problems/valid-parentheses/discuss/9178/Short-java-solution
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<> ();
        for (int i=0; i<s.length(); i++) {
            char c = s.charAt(i);
            if (c=='(') {
                stack.push(')');
            } else if (c=='{') {
                stack.push('}');
            } else if (c=='[') {
                stack.push(']');
            } else {
                if (stack.isEmpty() || stack.pop()!=c) {
                    return false;
                }
            }
        }
        return stack.isEmpty(); // consider more left parantheses than right parantheses cases
    }
}